File: javascript/lib/ApplicationControlExamples.js
/* 
 * Copyright 2019 Trend Micro.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

 /*
 * Modifies a policy to set the Application Control state to ON.
 * @param {ApiClient} api The Deep Security API exports.
 * @param {String} policyID The ID of the policy to modify.
 * @param {String} apiVersion The API version to use.
 * @returns {Promise} A promise object that resolves to the ID of the modified policy.
*/
exports.configureApplicationControl = function (api, policyID, apiVersion) {
  return new Promise((resolve, reject) => {
    const policy = new api.Policy();
    const policiesApi = new api.PoliciesApi();
    const appControlPolicyExtension = new api.ApplicationControlPolicyExtension();

    //Turn on application control
    appControlPolicyExtension.state =
      api.ApplicationControlPolicyExtension.StateEnum.on;

    //Add to the policy
    policy.applicationControl = appControlPolicyExtension;

    //Send the change to Deep Security Manager
    policiesApi.modifyPolicy(policyID, policy, apiVersion, { overrides: false })
      .then(data => {
        resolve(data);
      })
      .catch(function (error) {
        reject(error);
      });
  });
};

/*
 * Blocks all software changes on a computer.
 * @param {Number} computerID The ID of the computer.
 * @param {ApiClient} api The Deep Security API exports.
 * @param {String} apiVersion The api version to use.
 * @returns {Promise} A promise object that resolves to an Array of SoftwareChangeReviewResult objects.
 */
exports.blockAllUnrecognizedSoftware = function (computerID, api, apiVersion) {
  return new Promise((resolve, reject) => {
    // Search for software changes on the computer
    // Search criteria
    const computerCriteria = new api.SearchCriteria();
    computerCriteria.fieldName = "computerID";
    computerCriteria.numericTest = api.SearchCriteria.NumericTestEnum.equal;
    computerCriteria.numericValue = computerID;

    // Add criteria to search filter
    const searchFilter = new api.SearchFilter();
    searchFilter.searchCriteria = [computerCriteria];

    // Add search filter to a search options object
    const searchOptions = {
      searchFilter: searchFilter
    };

    // Perform the search
    const softwareChangeApi = new api.SoftwareChangesApi();
    softwareChangeApi.searchSoftwareChanges(apiVersion, searchOptions)
      .then(results => {
        // Block the unrecognized software

        // Create the software change review object and set action to block
        const softwareChangeReview = new api.SoftwareChangeReview();
        softwareChangeReview.action = api.SoftwareChangeReview.ActionEnum.block;

        if (results.softwareChanges.length > 0) {
          // Add the IDs of the software changes to block
          softwareChangeReview.softwareChangeIDs = results.softwareChanges.map(softwareChange => softwareChange.ID);

          // Perform the software change review
          softwareChangeApi.reviewSoftwareChanges(softwareChangeReview, apiVersion)
            .then(softwareChangeReview => {
              resolve(softwareChangeReview.softwareChangeReviewResults);
            })
            .catch(error => {
              reject(error);
            });
        } else {
          resolve(results);
        }
      })
      .catch(error => {
        reject(error);
      })
  });
};

/*
 * Creates a shared ruleset from a computer's software inventory
 * @param {Number} computerID The ID of the computer whose inventory the ruleset should be created from.
 * @param {String} rulesetName The name of the ruleset.
 * @param {ApiClient} api The Deep Security API exports.
 * @param {String} apiVersion The api version to use.
 * @returns {Promise} A promise object that resolves to a Ruleset object.
 */
exports.createSharedRuleset = function(computerID, rulesetName, api, apiVersion) {
  return new Promise((resolve, reject) => {
    // Create a software inventory
    const softwareInventory = new api.SoftwareInventory();
    softwareInventory.computerID = computerID;
    
    const softwareInventoryApi = new api.SoftwareInventoriesApi();
    softwareInventoryApi.createSoftwareInventory(softwareInventory, apiVersion)
      .then(softwareInventory => {
        // Wait until the software inventory is done building
        function waitForInventoryBuild() {
          softwareInventoryApi.describeSoftwareInventory(softwareInventory.ID, apiVersion)
            .then(softwareInventory => {
              if (softwareInventory.state === api.SoftwareInventory.StateEnum.complete) {
                // Create a ruleset
                const ruleset = new api.Ruleset();
                ruleset.name = rulesetName;

                const rulesetApi = new api.RulesetsApi();
                rulesetApi.createRuleset(ruleset, softwareInventory.ID, apiVersion)
                  .then(ruleset => {
                    return resolve(ruleset);
                  })
                  .catch(error => {
                    return reject(error);
                  });
              } else {
                //Check every 30 seconds
                setTimeout(waitForInventoryBuild, 30000);
              }
            })
            .catch(error => {
              return reject(error);
            });
        }
        waitForInventoryBuild();
      })
      .catch(error => {
        reject(error);
      });
  });
};

/*
 * Adds new Global Rules
 * @param {Array} sha256List The list of SHA-256 hashes of the executables to create new rules for.
 * @param {ApiClient} api The Deep Security API exports.
 * @param {String} apiVersion The api version to use.
 * @returns {Promise} A promise object that resolves to an Array of ApplicationControlGlobalRule objects.
 */
exports.addGlobalRules = function (sha256List, api, apiVersion) {
  return new Promise((resolve, reject) => {
    // Create the rules
    const globalRules = new api.ApplicationControlGlobalRules();
    globalRules.applicationControlGlobalRules = sha256List.map(sha256 => {
      const newRule = new api.ApplicationControlGlobalRule();
      newRule.sha256 = sha256;
      return newRule;
    });

    // Add the rules
    const globalRuleApi = new api.GlobalRulesApi();
    globalRuleApi.addGlobalRules(globalRules, apiVersion)
    .then(rules => {
      resolve(rules);
    })
    .catch(error => {
      reject(error);
    });
  });
};

/*
 * Turn on maintenance mode on a computer.
 * @param {Number} computerID The ID of the computer.
 * @param {Number} duration The maintenance mode duration in milliseconds.
 * @param {ApiClient} api The Deep Security API exports.
 * @param {String} apiVersion The api version to use.
 * @returns {Promise} A promise object that resolves to a Computer object. 
 */
exports.turnOnMaintenanceMode = function (computerID, duration, api, apiVersion) {
  // Create and configure an ApplicationControlComputerExtension
  const applicationControl = new api.ApplicationControlComputerExtension();
  applicationControl.maintenanceModeStatus = api.ApplicationControlComputerExtension.MaintenanceModeStatusEnum.on;
  applicationControl.maintenanceModeDuration = duration;

  // Add the ApplicationControlComputerExtension to a computer
  const computer = new api.Computer();
  computer.applicationControl = applicationControl;

  // Update the computer
  const computersApi = new api.ComputersApi();
  return computersApi.modifyComputer(computerID, computer, apiVersion);
}

File: python/src/application_control_examples.py
# Copyright 2019 Trend Micro.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

import time


def configure_application_control(api, configuration, api_version, api_exception, policy_id):
    """ Modifies a policy to set the application control state to on.

    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :param policy_id: The ID of the policy to modify.
    :return: A PoliciesApi object that contains the ID of the modified policy.
    """

    # Get the policy
    policies_api = api.PoliciesApi(api.ApiClient(configuration))

    # Turn on Application Control
    application_control_policy_extension = api.ApplicationControlPolicyExtension()
    application_control_policy_extension.state = "on"

    # Update the policy
    update_policy = api.Policy()
    update_policy.application_control = application_control_policy_extension

    # Modify the policy on Deep Security Manager
    app_control_policy = policies_api.modify_policy(policy_id, update_policy, api_version, overrides=False)
    return app_control_policy


def add_global_rules(sha256_list, api, configuration, api_version, api_exception):
    """ Adds new Global Rules

    :param sha256_list: The list of SHA-256 hashes of the executables to create new rules for.
    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :return: The list of new rules added.
    """

    # Create the rules
    new_rules = []
    for sha256 in sha256_list:
        new_rule = api.ApplicationControlGlobalRule()
        new_rule.sha256 = sha256
        new_rules.append(new_rule)

    # Add the rules
    global_rules_api = api.GlobalRulesApi(api.ApiClient(configuration))
    rules_list = api.ApplicationControlGlobalRules()
    rules_list.application_control_global_rules = new_rules
    return global_rules_api.add_global_rules(rules_list, api_version)


def block_all_unrecognized_software(computer_id, api, configuration, api_version, api_exception):
    """ Blocks all software changes on a computer.

    :param computer_id: The ID of the computer.
    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :return: A list of SoftwareChangeReviewResult objects.
    """

    # Search for software changes on the computer
    # Search criteria
    search_criteria = api.SearchCriteria()
    search_criteria.field_name = "computerID"
    search_criteria.numeric_test = "equal"
    search_criteria.numeric_value = computer_id

    # Add criteria to search filter
    search_filter = api.SearchFilter(None, [search_criteria])

    # Perform the search
    software_changes_api = api.SoftwareChangesApi(api.ApiClient(configuration))
    computer_software_changes = software_changes_api.search_software_changes(api_version, search_filter=search_filter)

    # Block the unrecognized software
    # Create the software change review object and set action to block
    software_change_review = api.SoftwareChangeReview()
    software_change_review.action = "block"
    software_change_review.software_change_ids = []

    # Add the IDs of the software changes to block
    for software_change in computer_software_changes.software_changes:
        software_change_review.software_change_ids.append(software_change.id)

    # Perform the software change review if software changes happened
    if len(software_change_review.software_change_ids) > 0:
        return software_changes_api.review_software_changes(software_change_review, api_version)


def create_shared_ruleset(computer_id, ruleset_name, api, configuration, api_version, api_exception):
    """ Creates a shared ruleset from a computer's software inventory.

    :param computer_id: The ID of the computer.
    :param ruleset_name: The name of the ruleset.
    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :return: A list of SoftwareChangeReviewResult objects.
    """
    software_inventory = api.SoftwareInventory()
    software_inventory.computer_id = computer_id

    # Build software_inventory
    software_inventories_api = api.SoftwareInventoriesApi(
        api.ApiClient(configuration))
    new_inventory = software_inventories_api.create_software_inventory(
        software_inventory, api_version)

    while new_inventory.state != "complete":
        # check status every 30 seconds
        time.sleep(30)
        new_inventory = software_inventories_api.describe_software_inventory(
            new_inventory.id, api_version)

    # Create ruleset
    ruleset = api.Ruleset()
    ruleset.name = ruleset_name
    rulesets_api = api.RulesetsApi(api.ApiClient(configuration))
    return rulesets_api.create_ruleset(ruleset, new_inventory.id, api_version)


def turn_on_maintenance_mode(computer_id, duration, api, configuration, api_version, api_exception):
    """ Turns on maintenance mode on a computer.

    :param computer_id: The ID of the computer.
    :param duration: The maintenance mode duration.
    :param api: The Deep Security API modules.
    :param configuration: Configuration object to pass to the api client.
    :param api_version: The version of the API to use.
    :param api_exception: The Deep Security API exception module.
    :return: A Computer object.
    """

    # Create and configure an ApplicationControlComputerExtesnion object
    application_control = api.ApplicationControlComputerExtension()
    application_control.maintenance_mode_status = "on"
    application_control.maintenance_mode_duration = duration

    # Add the ApplicationControlComputerExtension to a Computer object
    computer = api.Computer()
    computer.application_control = application_control

    # Update the computer
    computers_api = api.ComputersApi(api.ApiClient(configuration))
    return computers_api.modify_computer(computer_id, computer, api_version)
